# Lab | Pandas
---

- In this lab, we will be working with the customer data from an insurance company, 
    - which can be found in the CSV file located at the following link: https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv

The data includes information such as customer **ID, state, gender, education, income,** and **other variables** that can be used to perform various analyses.

- Throughout the lab, we will be using the pandas library in Python to manipulate and analyze the data. 
- Pandas is a powerful library that provides various data manipulation and analysis tools, 
    - including the ability to load and manipulate data from a variety of sources, including CSV files.

---
### Data Description

- Customer - Customer ID

- ST - State where customers live

- Gender - Gender of the customer

- Education - Background education of customers 

- Customer Lifetime Value - Customer lifetime value(CLV) is the total revenue the client will derive from their entire relationship with a customer. 
    - In other words, is the predicted or calculated value of a customer over their entire duration as a policyholder with the insurance company. 
    - It is an estimation of the net profit that the insurance company expects to generate from a customer throughout their relationship with the company. 
    - Customer Lifetime Value takes into account factors such as the duration of the customer's policy, premium payments, claim history, renewal likelihood, and potential additional services or products the customer may purchase. 
    - It helps insurers assess the long-term profitability and value associated with retaining a particular customer.

- Income - Customers income

- Monthly Premium Auto - Amount of money the customer pays on a monthly basis as a premium for their auto insurance coverage. 
    - It represents the recurring cost that the insured person must pay to maintain their insurance policy and receive coverage for potential damages, accidents, or other covered events related to their vehicle.

- Number of Open Complaints - Number of complaints the customer opened

- Policy Type - There are three type of policies in car insurance (Corporate Auto, Personal Auto, and Special Auto)

- Vehicle Class - Type of vehicle classes that customers have Two-Door Car, Four-Door Car SUV, Luxury SUV, Sports Car, and Luxury Car

- Total Claim Amount - the sum of all claims made by the customer. 
    - It represents the total monetary value of all approved claims for incidents such as accidents, theft, vandalism, or other covered events.


External Resources: https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9

---
---
---
# Solved Lab | Pandas
---

## Learning Goals
  Upon completion of this lab, you will be able to:
  
- Create Pandas DataFrames from CSV files.
- Explore Pandas DataFrames to get the **number of rows, columns, summary statistics and data types**.
- **Manipulate Series objects using various methods**, including `sort_values` or `value_counts`.
- **Access data in DataFrames through columns** (as a dictionary or attribute) and by **using** `loc` or `iloc` to **access rows**.
- **Filter** and **select data** in DataFrames **using simple comparison and logical operators**

---
## Introduction

- In this lab, you will learn how to work with and create pandas series and dataframes. 
- These are powerful tools for data manipulation and analysis in Python 
  - and allow you to create, manipulate, and analyze large, complex datasets with ease,*
    - making it a valuable library for any data scientist or analyst.

- You will also learn how to access their attributes and values, 
  - and manipulate them using a variety of built-in methods as well as learn how to filter and sort data and summarize and describe data. 
- **By the end of this lab**, you will have a strong foundation in working with **pandas series and dataframes, which you can apply to real-world data analysis problems**.

---
## About the dataset
### Context
- The dataset comprises customer data related to their vehicle insurance policies. 
- It provides details about both the customers and the specific insurance coverage they've opted for. This data an be explored to segment similar kinds of customers.

---

### Content
- This dataset provides socio-economic information of customers alongside details of their insured vehicles.
- The data contains both categorical and numerical variables. 
- The "Customer Lifetime Value" (CLV) has also been provided, which is derived from historical data.
- This is essential in understanding the customer purchase behavior.

---

### Inspiration

- The data is useful for clustering customers according to their behavior, thereby enabling targeted marketing of a new insurance policy. 
- This helps pre-estimating a budget for marketing expenses.


---
## DATA DESCRIPTION
| Data Field | Definition / Description |
|-----------|--------------------------|
| ``Customer`` | Customer ID |
| ``ST`` | State where the customer lives |
| ``Gender`` | Gender of the customer |
| ``Education`` | Educational background of the customer |
| ``Customer Lifetime Value`` | Customer Lifetime Value (CLV) is the total revenue the client will derive from their entire relationship with a customer. In other words, it is the predicted or calculated value of a customer over their entire duration as a policyholder with the insurance company. It is an estimation of the net profit that the insurance company expects to generate from a customer throughout their relationship with the company. CLV takes into account factors such as the duration of the customer's policy, premium payments, claim history, renewal likelihood, and potential additional services or products the customer may purchase. It helps insurers assess the long-term profitability and value associated with retaining a particular customer. |
| ``Income`` | Customer's earnings |
| ``Monthly Premium Auto`` | The monthly premium amount a customer pays for their auto insurance. It represents the recurring cost that the insured person must pay to maintain their insurance policy and receive coverage for potential damages, accidents, or other covered events related to their vehicle |
| ``Number of Open Complaints`` | Number of complaints the customer has opened |
| ``Policy Type`` | Insurance policy categories include Corporate Auto, Personal Auto, and Special Auto |
| ``Vehicle Class`` | Specifies the class of insured vehicles, such as Two-Door Car, Four-Door Car SUV, Luxury SUV, Sports Car, and Luxury Car |
| ``Total Claim Amount`` | The sum of all claims made by the customer. It represents the total monetary value of all approved claims for incidents such as accidents, theft, vandalism, or other covered events |


---
### EXPLORE

In [ ]:
import pandas as pd

df = pd.read_csv("https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv")

# Dimensions of the dataframe --------------------------------------------------
print("\nThe dimensions of the dataframe:", df.shape)

# Data types and non-null counts --------------------------------------------------
print("\nDataframe info:")
print(df.info())

# Summary statistics for numerical columns --------------------------------------------------
print("\nSummary statistics for numerical columns:")
numerical_summary = df.describe()
print(numerical_summary)

# Additional numerical insights --------------------------------------------------
print("\nMedian values for numerical columns:")
print(df.median(numeric_only=True))

print("\nMode values for numerical columns:")
print(df.mode(numeric_only=True).iloc[0])

# Summary statistics for categorical columns  --------------------------------------------------
print("\nSummary statistics for categorical columns:")
categorical_summary = df.describe(include='object')
print(categorical_summary)

# Value counts for main categorical variables  --------------------------------------------------
print("\nPolicy Type distribution:")
print(df['Policy Type'].value_counts())

print("\nState (ST) distribution:")
print(df['ST'].value_counts())

print("\nGender distribution:")
print(df['GENDER'].value_counts())

print("\nEducation distribution:")
print(df['Education'].value_counts())



The dimensions of the dataframe: (4008, 11)

Dataframe info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4008 entries, 0 to 4007
Data columns (total 11 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Customer                   1071 non-null   object 
 1   ST                         1071 non-null   object 
 2   GENDER                     954 non-null    object 
 3   Education                  1071 non-null   object 
 4   Customer Lifetime Value    1068 non-null   object 
 5   Income                     1071 non-null   float64
 6   Monthly Premium Auto       1071 non-null   float64
 7   Number of Open Complaints  1071 non-null   object 
 8   Policy Type                1071 non-null   object 
 9   Vehicle Class              1071 non-null   object 
 10  Total Claim Amount         1071 non-null   float64
dtypes: float64(3), object(8)
memory usage: 344.6+ KB
None

Summary statistics for numerical columns:
 

In [ ]:
# Your code here
import pandas as pd
df = pd.read_csv("https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv")

# dimensions of the dataframe
print("\nThe dimensions of the dataframe:", df.shape)

# data types of each column
print(df.info())

# evaluation variable nature ---------------------------------------------------------------------

# summary statistics for numerical columns -------------------------------------------------------
# to understand the central tendency and distribution of the data
# mean, median, mode, standard deviation, quartiles
# conclusions based on summaries statistics

# summary statistics for categorical columns -----------------------------------------------------
# conclusions based on summaries statistics

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4008 entries, 0 to 4007
Data columns (total 11 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Customer                   1071 non-null   object 
 1   ST                         1071 non-null   object 
 2   GENDER                     954 non-null    object 
 3   Education                  1071 non-null   object 
 4   Customer Lifetime Value    1068 non-null   object 
 5   Income                     1071 non-null   float64
 6   Monthly Premium Auto       1071 non-null   float64
 7   Number of Open Complaints  1071 non-null   object 
 8   Policy Type                1071 non-null   object 
 9   Vehicle Class              1071 non-null   object 
 10  Total Claim Amount         1071 non-null   float64
dtypes: float64(3), object(8)
memory usage: 344.6+ KB
None

The dimensions of the dataframe: (4008, 11)


---
## Challenge 1: Understanding the data

In this challenge, you will use pandas to explore a given dataset. 

Your task is to gain a **deep understanding of the data** by **analyzing its** ``characteristics``, ``dimensions``, and ``statistical properties``.

1. Load CSV in a DataFrame
2. Identify the **dimensions of the dataset** by determining the **number of rows and columns** it contains.
3. Determine the **data types of each column** and **evaluate whether they are appropriate for the nature of the variable**. 
    - You should also **provide suggestions for fixing any incorrect data types**.
4. Identify the **number of unique values for each column** and **determine which columns appear to be categorical**. 
    - You should also **describe the unique values of each categorical column** and **the range of values for numerical columns**, and **give your insights**.
5. Compute **summary statistics** such as ``mean``, ``median``, ``mode``, ``standard deviation``, and ``quartiles`` 
    - to **understand the central tendency and distribution of the data for** ``numerical columns``. 
    - You should also **provide your conclusions based on these summary statistics**.
6. Compute **summary statistics** for **categorical columns** and **providing your conclusions based on these statistics**.

---
### 1. Load CSV in a DataFrame

In [1]:
import pandas as pd
import numpy as np

# Load the dataset
df = pd.read_csv("https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv")

---
### 2. Identify the **dimensions of the dataset** by determining the **number of rows and columns** it contains.

In [2]:
## 1. Dimensions
print("\nDataset Dimensions:")

# df.shape returns a tuple (rows, columns)
# df.shape[0] gives number of rows
# df.shape[1] gives number of columns
print(f"Rows: {df.shape[0]}")
print(f"Columns: {df.shape[1]}")
print(f"shape of this dataset:", df.shape)
print(f"View first 5 rows of the dataset:")
df.head(20)


Dataset Dimensions:
Rows: 4008
Columns: 11
shape of this dataset: (4008, 11)
View first 5 rows of the dataset:


Customer          ST GENDER             Education Customer Lifetime Value  \
0   RB50392  Washington    NaN                Master                     NaN   
1   QZ44356     Arizona      F              Bachelor              697953.59%   
2   AI49188      Nevada      F              Bachelor             1288743.17%   
3   WW63253  California      M              Bachelor              764586.18%   
4   GA49547  Washington      M  High School or Below              536307.65%   
5   OC83172      Oregon      F              Bachelor              825629.78%   
6   XZ87318      Oregon      F               College              538089.86%   
7   CF85061     Arizona      M                Master              721610.03%   
8   DY87989      Oregon      M              Bachelor             2412750.40%   
9   BQ94931      Oregon      F               College              738817.81%   
10  SX51350  California      M               College              473899.20%   
11  VQ65197  California    NaN               College              819719.71%   
12  DP39365  California    NaN                Master              879879.70%   
13  SJ95423     Arizona    NaN  High School or Below              881901.89%   
14  IL66569  California    NaN               College              538443.17%   
15  BW63560      Oregon    NaN              Bachelor              746313.94%   
16  FV94802      Nevada    NaN  High School or Below              256686.78%   
17  OE15005        Cali    NaN               College              394524.16%   
18  WC83389      Oregon    NaN               College              571033.31%   
19  FL50705  California    NaN  High School or Below              816261.71%   

     Income  Monthly Premium Auto Number of Open Complaints     Policy Type  \
0       0.0                1000.0                    1/0/00   Personal Auto   
1       0.0                  94.0                    1/0/00   Personal Auto   
2   48767.0                 108.0                    1/0/00   Personal Auto   
3       0.0                 106.0                    1/0/00  Corporate Auto   
4   36357.0                  68.0                    1/0/00   Personal Auto   
5   62902.0                  69.0                    1/0/00   Personal Auto   
6   55350.0                  67.0                    1/0/00  Corporate Auto   
7       0.0                 101.0                    1/0/00  Corporate Auto   
8   14072.0                  71.0                    1/0/00  Corporate Auto   
9   28812.0                  93.0                    1/0/00    Special Auto   
10      0.0                  67.0                    1/0/00   Personal Auto   
11      0.0                 110.0                    1/0/00   Personal Auto   
12  77026.0                 110.0                    1/2/00  Corporate Auto   
13  99845.0                 110.0                    1/1/00  Corporate Auto   
14  83689.0                  70.0                    1/2/00  Corporate Auto   
15  24599.0                  64.0                    1/1/00  Corporate Auto   
16  25049.0                  67.0                    1/0/00   Personal Auto   
17  28855.0                 101.0                    1/0/00   Personal Auto   
18  51148.0                  72.0                    1/0/00   Personal Auto   
19  66140.0                 101.0                    1/0/00  Corporate Auto   

    Vehicle Class  Total Claim Amount  
0   Four-Door Car            2.704934  
1   Four-Door Car         1131.464935  
2    Two-Door Car          566.472247  
3             SUV          529.881344  
4   Four-Door Car           17.269323  
5    Two-Door Car          159.383042  
6   Four-Door Car          321.600000  
7   Four-Door Car          363.029680  
8   Four-Door Car          511.200000  
9   Four-Door Car          425.527834  
10  Four-Door Car          482.400000  
11            SUV          528.000000  
12  Four-Door Car          472.029737  
13            SUV          528.000000  
14  Four-Door Car          307.139132  
15  Four-Door Car     

---
### 3. Determine the **data types of each column** and **evaluate whether they are appropriate for the nature of the variable**. 
- You should also **provide suggestions for fixing any incorrect data types**.

In [3]:
## 2. Data Types and Appropriateness

# Display data types of each column without non-nulls count
print("\nData Types of each column:")
print(df.dtypes)

# Display data types of each column with non-nulls count
print("\nDataframe Info:")
print(df.info())

# df['column_name'] = df['column_name'].astype(new_type)
print("\nSuggestions to fix incorrect data types:")

print("- Customer Lifetime Value: object to float (remove '%', divide by 100)")
print("- GENDER: Standardize variants mapping (Male/male/M to 'M')")
print("- ST: Standardize state codes mapping (Cali to California, WA to Washington)")
print("- Number of Open Complaints: object to int (parse '1/0/00' format)")



Data Types of each column:
Customer                      object
ST                            object
GENDER                        object
Education                     object
Customer Lifetime Value       object
Income                       float64
Monthly Premium Auto         float64
Number of Open Complaints     object
Policy Type                   object
Vehicle Class                 object
Total Claim Amount           float64
dtype: object

Dataframe Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4008 entries, 0 to 4007
Data columns (total 11 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Customer                   1071 non-null   object 
 1   ST                         1071 non-null   object 
 2   GENDER                     954 non-null    object 
 3   Education                  1071 non-null   object 
 4   Customer Lifetime Value    1068 non-null   object 
 5   Income                     107

---
#### clean Customer Lifetime Value

In [4]:
print("Before:", df['Customer Lifetime Value'].head(10))
df['Customer Lifetime Value'] = df['Customer Lifetime Value'].str.replace('%', '').astype(float) / 100
print("After:", df['Customer Lifetime Value'].head(10))

Before: 0            NaN
1     697953.59%
2    1288743.17%
3     764586.18%
4     536307.65%
5     825629.78%
6     538089.86%
7     721610.03%
8    2412750.40%
9     738817.81%
Name: Customer Lifetime Value, dtype: object
After: 0           NaN
1     6979.5359
2    12887.4317
3     7645.8618
4     5363.0765
5     8256.2978
6     5380.8986
7     7216.1003
8    24127.5040
9     7388.1781
Name: Customer Lifetime Value, dtype: float64


---
#### clean GENDER

In [5]:
print("Before GENDER:", df['GENDER'].unique())
# mapping GENDER values
df['GENDER'] = df['GENDER'].map({
    'F': 'F',
    'Femal': 'F',
    'female': 'F',
    'M': 'M',
    'Male': 'M'
})
print("After GENDER:", df['GENDER'].unique())


Before GENDER: [nan 'F' 'M' 'Femal' 'Male' 'female']
After GENDER: [nan 'F' 'M']


---
#### clean ST

In [8]:
print("Before ST:", df['ST'].unique())
# mapping ST values
df['ST'] = df['ST'].map({
    'WA': 'Washington',
    'AZ': 'Arizona',
    'Cali': 'California'
})
print("After ST:", df['ST'].unique())

Before ST: [nan]
After ST: [nan]


---
#### Number of Open Complaints

In [10]:
print("Before:", df['Number of Open Complaints'].head(10))

df['Number of Open Complaints'] = pd.to_numeric(
    df['Number of Open Complaints'].str.split('/').str[0],
    errors='coerce'
).fillna(0).astype(int)

print("After:", df['Number of Open Complaints'].head(10))


Before: 0    1/0/00
1    1/0/00
2    1/0/00
3    1/0/00
4    1/0/00
5    1/0/00
6    1/0/00
7    1/0/00
8    1/0/00
9    1/0/00
Name: Number of Open Complaints, dtype: object
After: 0    1
1    1
2    1
3    1
4    1
5    1
6    1
7    1
8    1
9    1
Name: Number of Open Complaints, dtype: int64


---
### 4. Identify the **number of unique values for each column** and **determine which columns appear to be categorical**. 
- You should also **describe** the `unique values` of `each categorical column` **and** the `range of values` **for** `numerical columns`, and `**give your insights**`.

In [ ]:
# 3. Unique Values & Categorical Identification
print("\nUnique Values per Column:")
for col in df.columns:
    print(f"{col} has {df[col].nunique()} unique values")

# Identify categorical columns (unique < 10) and handle empty columns NaN
categorical_cols = [col for col in df.columns if df[col].nunique() < 10 and df[col].nunique() > 0]
print(f"\nCategorical Columns (unique < 10 and not empty): \n{categorical_cols}")

for col in categorical_cols:
    print(f"\nCategorical column: {col}")
    print(df[col].value_counts().head())

# Identify numerical columns
numerical_cols = [col for col in df.columns if pd.api.types.is_numeric_dtype(df[col]) and col not in categorical_cols]
for col in numerical_cols:
    print(f"\n{col} numerical column range: {df[col].min()} to {df[col].max()}")

# Insights
print("\nInsights:")

## Categorical column insights
for col in categorical_cols:
    if not df[col].mode().empty:
        print(f"{col}: Most common value is '{df[col].mode()[0]}'")
    else:
        print(f"{col}: No mode available")

## Numerical column insights
for col in numerical_cols:
    print(f"{col}: Average is {df[col].mean():.2f}, Median is {df[col].median():.2f}")



Unique Values per Column:
Customer has 1071 unique values
ST has 0 unique values
GENDER has 2 unique values
Education has 6 unique values
Customer Lifetime Value has 1027 unique values
Income has 774 unique values
Monthly Premium Auto has 132 unique values
Number of Open Complaints has 2 unique values
Policy Type has 3 unique values
Vehicle Class has 6 unique values
Total Claim Amount has 761 unique values

Categorical Columns (unique < 10 and not empty): 
['GENDER', 'Education', 'Number of Open Complaints', 'Policy Type', 'Vehicle Class']

Categorical column: GENDER
GENDER
F    502
M    452
Name: count, dtype: int64

Categorical column: Education
Education
Bachelor                324
College                 313
High School or Below    296
Master                   94
Doctor                   37
Name: count, dtype: int64

Categorical column: Number of Open Complaints
Number of Open Complaints
0    2937
1    1071
Name: count, dtype: int64

Categorical column: Policy Type
Policy Type
Per

In [21]:
## 4. Numerical Summary Statistics
print("\nNumerical Summary Statistics:")
numerical_cols = df.select_dtypes(include=[np.number]).columns
print(df[numerical_cols].describe().round())

print("\nAdditional Numerical Stats:")
print("Medians:", df[numerical_cols].median().round())
print("Modes:", df[numerical_cols].mode().iloc[0].round())


Numerical Summary Statistics:
       Customer Lifetime Value   Income  Monthly Premium Auto  \
count                   1068.0   1071.0                1071.0   
mean                    7937.0  39296.0                 193.0   
std                     6435.0  30469.0                1601.0   
min                     2004.0      0.0                  61.0   
25%                     4034.0  14072.0                  68.0   
50%                     5882.0  36234.0                  83.0   
75%                     8963.0  64631.0                 110.0   
max                    58167.0  99960.0               35354.0   

       Number of Open Complaints  Total Claim Amount  
count                     4008.0              1071.0  
mean                         0.0               405.0  
std                          0.0               293.0  
min                          0.0                 0.0  
25%                          0.0               202.0  
50%                          0.0               355.0 

---
### 5. Compute **summary statistics** such as ``mean``, ``median``, ``mode``, ``standard deviation``, and ``quartiles`` 
- to **understand the central tendency and distribution of the data for** ``numerical columns``. 
- You should also **provide your conclusions based on these summary statistics**.

In [25]:
# Select numerical columns
numerical_cols = df.select_dtypes(include=['int64', 'float64']).columns

# Compute summary statistics
summary_stats = df[numerical_cols].describe().T.round(2)  # mean, std, min, 25%, 50%, 75%, max
summary_stats['mode'] = df[numerical_cols].mode().iloc[0].round(2)  # add mode

print("Summary Statistics for Numerical Columns:\n")
print(summary_stats)

# Optional: Conclusions / Insights
print("\nConclusions / Insights:")
for col in numerical_cols:
    mean = summary_stats.loc[col, 'mean'].round(2)
    median = summary_stats.loc[col, '50%'].round(2)
    mode = summary_stats.loc[col, 'mode'].round(2)
    std = summary_stats.loc[col, 'std'].round(2)
    q1 = summary_stats.loc[col, '25%'].round(2)
    q3 = summary_stats.loc[col, '75%'].round(2)
    
    print(f"\nColumn: {col}")
    print(f"  - Mean: {mean:.2f}")
    print(f"  - Median: {median}")
    print(f"  - Mode: {mode}")
    print(f"  - Standard Deviation: {std:.2f}")
    print(f"  - Quartiles: Q1={q1}, Q3={q3}")
    
    # Basic interpretation
    if std > mean*0.5:
        print("  -> Data is widely spread / high variability")
    else:
        print("  -> Data is relatively concentrated / low variability")
    
    if abs(mean - median) > std*0.5:
        print("  -> Data may be skewed")
    else:
        print("  -> Data is relatively symmetric")


Summary Statistics for Numerical Columns:

                            count      mean       std      min       25%  \
Customer Lifetime Value    1068.0   7936.90   6434.78  2004.35   4034.08   
Income                     1071.0  39295.70  30469.43     0.00  14072.00   
Monthly Premium Auto       1071.0    193.23   1601.19    61.00     68.00   
Number of Open Complaints  4008.0      0.27      0.44     0.00      0.00   
Total Claim Amount         1071.0    404.99    293.03     0.38    202.16   

                                50%       75%       max     mode  
Customer Lifetime Value     5881.74   8962.87  58166.55  2514.59  
Income                     36234.00  64631.00  99960.00     0.00  
Monthly Premium Auto          83.00    109.50  35354.00    65.00  
Number of Open Complaints      0.00      1.00      1.00     0.00  
Total Claim Amount           354.73    532.80   2893.24   321.60  

Conclusions / Insights:

Column: Customer Lifetime Value
  - Mean: 7936.90
  - Median: 5881.74
  

#### Summary Statistics & Conclusions

#### Customer Lifetime Value
| Statistic           | Value       |
|--------------------|------------|
| Mean               | 7936.90    |
| Median             | 5881.74    |
| Mode               | 2514.59    |
| Standard Deviation | 6434.78    |
| Q1                 | 4034.08    |
| Q3                 | 8962.87    |
| Conclusion         | Data is widely spread with high variability. Distribution is relatively symmetric. |

#### Income
| Statistic           | Value       |
|--------------------|------------|
| Mean               | 39295.70   |
| Median             | 36234.00   |
| Mode               | 0.00       |
| Standard Deviation | 30469.43   |
| Q1                 | 14072.00   |
| Q3                 | 64631.00   |
| Conclusion         | Data is widely spread / high variability. Distribution is relatively symmetric, but presence of zero income indicates possible outliers or missing values. |

#### Monthly Premium Auto
| Statistic           | Value       |
|--------------------|------------|
| Mean               | 193.23     |
| Median             | 83.00      |
| Mode               | 65.00      |
| Standard Deviation | 1601.19    |
| Q1                 | 68.00      |
| Q3                 | 109.50     |
| Conclusion         | High variability due to extreme values (max=35,354). Most data is concentrated at lower premiums. Distribution is positively skewed. |

#### Number of Open Complaints
| Statistic           | Value       |
|--------------------|------------|
| Mean               | 0.27       |
| Median             | 0.00       |
| Mode               | 0.00       |
| Standard Deviation | 0.44       |
| Q1                 | 0.00       |
| Q3                 | 1.00       |
| Conclusion         | Most customers have no complaints; distribution is skewed toward 0. Low variability overall. |

#### Total Claim Amount
| Statistic           | Value       |
|--------------------|------------|
| Mean               | 404.99     |
| Median             | 354.73     |
| Mode               | 321.60     |
| Standard Deviation | 293.03     |
| Q1                 | 202.16     |
| Q3                 | 532.80     |
| Conclusion         | Data is moderately spread / high variability. Distribution is fairly symmetric. |


---
### 6. Compute **summary statistics** for **categorical columns** and **providing your conclusions based on these statistics**.

In [29]:
## 6. Categorical Distributions & Conclusions
# Select categorical columns (unique < 10 and not empty)
categorical_cols = [col for col in df.columns if df[col].nunique() < 10 and df[col].nunique() > 0]

print("### Summary Statistics & Conclusions for Categorical Columns\n")

for col in categorical_cols:
    counts = df[col].value_counts()
    mode = counts.idxmax()
    
    print(f"#### {col}")
    print(counts.to_frame(name='Count'))
    print(f"Mode: {mode}")
    


### Summary Statistics & Conclusions for Categorical Columns

#### GENDER
        Count
GENDER       
F         502
M         452
Mode: F
#### Education
                      Count
Education                  
Bachelor                324
College                 313
High School or Below    296
Master                   94
Doctor                   37
Bachelors                 7
Mode: Bachelor
#### Number of Open Complaints
                           Count
Number of Open Complaints       
0                           2937
1                           1071
Mode: 0
#### Policy Type
                Count
Policy Type          
Personal Auto     780
Corporate Auto    234
Special Auto       57
Mode: Personal Auto
#### Vehicle Class
               Count
Vehicle Class       
Four-Door Car    576
Two-Door Car     205
SUV              199
Sports Car        57
Luxury SUV        20
Luxury Car        14
Mode: Four-Door Car


#### **Conclusions**
- **Mode: F** 
    - Conclusion: Majority female, fairly balanced distribution.
- **Mode: Bachelor**  
    - Conclusion: Diverse education levels, Bachelor most common.
- **Mode: 0**
    - Conclusion: Most customers have no complaints.
- **Mode: Personal Auto**
    - Conclusion: Majority hold Personal Auto policies.
- **Mode: Four-Door Car**
    - Conclusion: Four-Door Car most common, luxury and sports cars are rare.

---
## Challenge 2: analyzing the data

---
### Exercise 1
The marketing team wants to know ```the top 5 less common customer locations```. 
- **Create** a pandas **Series** object that contains the `customer locations and their frequencies`, 
- and then **retrieve** the `top 5 less common locations in ascending order`.

In [9]:
import pandas as pd

# Load the dataset
url = "https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv"
df = pd.read_csv(url)

# Frequency of each customer location (state)
location_counts = df['ST'].value_counts()
print("Frequency of each customer location (state):\n",location_counts)
print("Type of location_counts:", type(location_counts))

# The 5 less common customer locations (smallest counts)
least_common_locations = location_counts.nsmallest(5)
print("\nThe 5 less common customer locations (smallest counts):\n",least_common_locations)
print("Type of least_common_locations:", type(least_common_locations))

# Display the result in ascending order
least_common_locations_sorted = least_common_locations.sort_values()
print("\nDisplay the result in ascending order:\n",least_common_locations_sorted)
print("Type of least_common_locations_sorted:", type(least_common_locations_sorted))


Frequency of each customer location (state):
 ST
Oregon        320
California    211
Arizona       186
Cali          120
Nevada         98
Washington     81
WA             30
AZ             25
Name: count, dtype: int64
Type of location_counts: <class 'pandas.core.series.Series'>

The 5 less common customer locations (smallest counts):
 ST
AZ             25
WA             30
Washington     81
Nevada         98
Cali          120
Name: count, dtype: int64
Type of least_common_locations: <class 'pandas.core.series.Series'>

Display the result in ascending order:
 ST
AZ             25
WA             30
Washington     81
Nevada         98
Cali          120
Name: count, dtype: int64
Type of least_common_locations_sorted: <class 'pandas.core.series.Series'>


---
### Exercise 2

The sales team wants to know ``the total number of policies sold`` **for** ``each type of policy``. 
- **Create** a ``pandas Series object`` that **contains** the ``policy types`` **and** ``their total number of policies sold``, and **then retrieve** the ``policy type with the highest number of policies sold``.

*Hint:*
- *Using ```value_counts() method``` simplifies this analysis.*
- *Futhermore, there is a ```method that returns the index of the maximum value in a column or row```.*


In [ ]:
import pandas as pd

# Load the dataset
url = "https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv"
df = pd.read_csv(url)

# Count the total number of policies sold for each policy type
policy_counts = df['Policy Type'].value_counts()

# Retrieve the policy type with the highest number of policies sold
top_policy = policy_counts.idxmax()  # returns the policy type
top_policy_count = policy_counts.max()  # returns the count

# Display results
print("Total number of policies sold by type:")
print(policy_counts)
print(f"\nPolicy type with the highest number of policies sold: {top_policy} ({top_policy_count} policies)")


Total number of policies sold by type:
Policy Type
Personal Auto     780
Corporate Auto    234
Special Auto       57
Name: count, dtype: int64

Policy type with the highest number of policies sold: Personal Auto (780 policies)


---
### Exercise 3

The sales team wants to know ```if customers with Personal Auto have a lower income than those with Corporate Auto```. 
- **How does the average income compare between the two policy types?**
- Use `loc` to **create two dataframes**: 
    - one containing only `Personal Auto policies` 
    - and one containing only `Corporate Auto policies`.
- Calculate the average income for each policy.
- Print the results.

In [4]:
import pandas as pd

# Load the dataset
url = "https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv"
df = pd.read_csv(url)

# Step 1: Create two DataFrames using .loc
personal_auto_df = df.loc[df['Policy Type'] == 'Personal Auto']
corporate_auto_df = df.loc[df['Policy Type'] == 'Corporate Auto']

# Step 2: Calculate the average income for each policy type
personal_auto_avg_income = personal_auto_df['Income'].mean()
corporate_auto_avg_income = corporate_auto_df['Income'].mean()

# Step 3: Print the results
print(f"Average income for Personal Auto customers: {personal_auto_avg_income:.2f}")
print(f"Average income for Corporate Auto customers: {corporate_auto_avg_income:.2f}")


Average income for Personal Auto customers: 38180.70
Average income for Corporate Auto customers: 41390.31


---
### Bonus: Exercise 4


#### Your **goal** is to **`identify customers with a high policy claim amount`**.

#### Instructions:

- Review again the statistics for total claim amount to gain an understanding of the data.
- To identify potential areas for improving customer retention and profitability, we want to focus on customers with a high policy claim amount. Consider customers with a high policy claim amount to be those in the top 25% of the total claim amount. Create a pandas DataFrame object that contains information about customers with a policy claim amount greater than the 75th percentile.
- Use DataFrame methods to calculate summary statistics about the high policy claim amount data. 

#### **Note:** 
- When analyzing data, we often want to **focus on certain groups of values to gain insights**. 
- **Percentiles are a useful tool to help us define these groups**. 
    - A percentile is a measure that tells us what percentage of values in a dataset are below a certain value. 
    - For example, the **75th percentile represents the value below which 75% of the data falls.** 
    - Similarly, the **25th percentile represents the value below which 25% of the data falls.** 
        - When we talk about the **`top 25%`**, we are **referring to the values that fall above the 75th percentile**, which `represent the top quarter of the data`. 
        - On the other hand, when we talk about the **`bottom 25%`**, we are referring to the values that `fall below the 25th percentile`, which `represent the bottom quarter of the data`. 
- By **`focusing on these groups`**, we can **`identify patterns and trends`** that may be **`useful for making decisions and taking action`**.

- **Hint: look for a method that gives you the percentile or quantile 0.75 and 0.25 for a Pandas Series.**
- **Hint 2: check `Boolean selection according to the values of a single column` in https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9**

In [ ]:
import pandas as pd

# Load the dataset
url = "https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv"
df = pd.read_csv(url)

# Total Claim Amount statistics -----------------------------------
print("Summary statistics for Total Claim Amount:")
print(df['Total Claim Amount'])
df['Total Claim Amount'].describe()

# Identifying the 75th percentile (top 25%) -----------------------------------
claim_75th_percentile = df['Total Claim Amount'].quantile(0.75)

print(f"\n75th percentile of Total Claim Amount: {claim_75th_percentile}")

# Create a DataFrame with high policy claim amount customers -----------------------------------
# customers above 75%
high_claim_df = df[df['Total Claim Amount'] > claim_75th_percentile]
print(f"\nNumber of customers in the top 25% of Total Claim Amount: {high_claim_df.shape[0]}")

# Calculate summary statistics for high claim customers -----------------------------------
print("\nSummary statistics for high policy claim amount customers:")
high_claim_df.describe()

Summary statistics for Total Claim Amount:
0          2.704934
1       1131.464935
2        566.472247
3        529.881344
4         17.269323
           ...     
4003            NaN
4004            NaN
4005            NaN
4006            NaN
4007            NaN
Name: Total Claim Amount, Length: 4008, dtype: float64

75th percentile of Total Claim Amount: 532.8

Number of customers in the top 25% of Total Claim Amount: 264

Summary statistics for high policy claim amount customers:


Income  Monthly Premium Auto  Total Claim Amount
count    264.000000            264.000000          264.000000
mean   23677.344697            165.193182          782.228263
std    27013.483721            623.930992          292.751640
min        0.000000             63.000000          537.600000
25%        0.000000             99.000000          606.521741
50%    18807.000000            114.000000          679.597985
75%    42423.750000            133.250000          851.400000
max    99316.000000          10202.000000         2893.239678